File: 0678-valid-parenthesis-string/solution.py
class Solution(object):
    def checkValidString(self, s):
        """
        :type s: str
        :rtype: bool
        """
        lmin=0
        lmax=0
        for i in s:
            if i=='(':
                lmin+=1
                lmax+=1
            elif i==')':
                lmin-=1
                lmax-=1
            elif i=='*':
                lmin-=1
                lmax+=1
            if lmax<0:
                return False
            elif lmin<0:
                lmin=0

        return lmin==0

File: 0678-valid-parenthesis-string/solution.java
class Solution {
    public boolean checkValidString(String s) {
        Stack<Integer> openStack = new Stack<>();
        Stack<Integer> starStack = new Stack<>();
        
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '(') {
                openStack.push(i);
            } else if (c == '*') {
                starStack.push(i);
            } else if (c == ')') {
                if (!openStack.isEmpty()) {
                    openStack.pop();
                } else if (!starStack.isEmpty()) {
                    starStack.pop();
                } else {
                    return false;
                }
            }
        }
        
        // Match remaining '(' with '*'
        while (!openStack.isEmpty() && !starStack.isEmpty()) {
            if (openStack.peek() < starStack.peek()) {
                openStack.pop();
                starStack.pop();
            } else {
                return false;
            }
        }
        
        // If there are unmatched '(' left, return false
        return openStack.isEmpty();
    }
}

